### PR TITLE
Add Model as defined on a class

### PR DESCRIPTION
**Pull request by canyon289**
Anyone know what the if method code block does is intended for below? None of the tests hit that entire code block

### CODE REVIEW

File: pymc4/coroutine_model.py [hunk 1]
```diff
@@ -15,6 +15,7 @@
 def model(genfn, *, name=_no_name_provided, keep_auxiliary=True, keep_return=True, method=False):
     """Flexibly wrap a generator function into a Model template."""
     if method:
+        # What is this block for?
```
Comment (canyon289) [inline pymc4/coroutine_model.py]: This one code block :)

Comment (ferrine) [reply pymc4/coroutine_model.py]: In case first argument is self

Comment (ferrine) [reply pymc4/coroutine_model.py]: And if you do it another way, the defined method is not recognised as function, it is recognised as an object and self is not passed

Comment (ferrine) [reply pymc4/coroutine_model.py]: And if you do it another way, the defined method is not recognised as function, it is recognised as an object and self is not passed

Comment (ferrine) [reply pymc4/coroutine_model.py]: And if you do it another way, the defined method is not recognised as function, it is recognised as an object and self is not passed

Comment (canyon289) [reply pymc4/coroutine_model.py]: Thanks

File: tests/test_executor.py [hunk 2]
```diff
@@ -66,6 +66,25 @@ def transformed_model_with_observed():
...
+    class PyMC4ClassModel:
+        @pm.model
+        def class_model_method(self):
+            norm = yield pm.Normal("n", 0, 1)
+            return norm
+
+    return PyMC4ClassModel()
+
+
+def test_class_model(class_model):
```
Comment (canyon289) [inline tests/test_executor.py]: Is this what it's supposed to look like? It works but its pretty ugly imo. Am I missing the more "pythonic" version?

**Comment by canyon289** [comment]
Thanks @ferrine Ready for review and merge then

**Comment by junpenglao** [comment]
I might have missed the discussion - what are we trying to do here? Is it to add support for any callable? to pm.sample (e.g., passing a class that has a `__call__` method)?

**Comment by canyon289** [comment]
@ferrine has written logic that allows for PyMC4 models to be written as methods on a class.

I'm trying to understand all of PyMC4 by writing tests for everything and asking @ferrine questions

**Comment by junpenglao** [comment]
Ah I see. In that case, could we create an arbitrary class in the test and monkey patch one of its method as a pymc4 model? In this case we can reuse the pymc4 models that we already defined in the same test.

**Comment by canyon289** [comment]
> Ah I see. In that case, could we create an arbitrary class in the test and monkey patch one of its method as a pymc4 model? In this case we can reuse the pymc4 models that we already defined in the same test.

Now I am confused. The class, and instantiated object is in a fixture so it can be reused, and since this is testing user facing API we shouldnt need monkeypatch, after all the user won't be monkeypatching.

**Comment by junpenglao** [comment]
Monkey patching just a way to save a few lines of defining the model in method again. But you are right, probably a more meaningful test case to spell out everything explicitly.

**Comment by canyon289** [comment]
Ah I see what you mean. I think its better to leave it explicit for now so there's an example of class based API for models.

If this looks good can I merge?

**Comment by junpenglao** [comment]
Thanks!

**Comment by ferrine** [comment]
The main usecase of defining model within a method is being able to interchange model parts by subclassing.

Here is a draft of how it might look like. `other_method` is easy to override, but copy-pasting the whole model may be error prone 
```python
class BaseModel:
    @model
    def define_my_model(self, ...):
        var = yield self.other_method(...)

    def other_method(self, ...):
        raise NotImplementedError

class ModelType1(BaseModel):
    def other_method(self, ...):
        return (yield pm.Normal(...))
```

**Pull request merged**